Mathlib/Data/Matrix/Defs.lean

```diff
@@ -0,0 +1,842 @@
+    (-A) i j = -(A i j) := rfl
+
+end
+
+/-! simp-normal form pulls `of` to the outside. -/
+
+@[simp]
+theorem of_zero [Zero α] : of (0 : m → n → α) = 0 :=
+  rfl
+
+@[simp]
+theorem of_add_of [Add α] (f g : m → n → α) : of f + of g = of (f + g) :=
+  rfl
+
+@[simp]
+theorem of_sub_of [Sub α] (f g : m → n → α) : of f - of g = of (f - g) :=
+  rfl
+
+@[simp]
+theorem neg_of [Neg α] (f : m → n → α) : -of f = of (-f) :=
+  rfl
+
+@[simp]
+theorem smul_of [SMul R α] (r : R) (f : m → n → α) : r • of f = of (r • f) :=
+  rfl
+
+@[simp]
+protected theorem map_zero [Zero α] [Zero β] (f : α → β) (h : f 0 = 0) :
+    (0 : Matrix m n α).map f = 0 := by
+  ext
+  simp [h]
+
+protected theorem map_add [Add α] [Add β] (f : α → β) (hf : ∀ a₁ a₂, f (a₁ + a₂) = f a₁ + f a₂)
+    (M N : Matrix m n α) : (M + N).map f = M.map f + N.map f :=
+  ext fun _ _ => hf _ _
+
+protected theorem map_sub [Sub α] [Sub β] (f : α → β) (hf : ∀ a₁ a₂, f (a₁ - a₂) = f a₁ - f a₂)
+    (M N : Matrix m n α) : (M - N).map f = M.map f - N.map f :=
+  ext fun _ _ => hf _ _
+
+theorem map_smul [SMul R α] [SMul R β] (f : α → β) (r : R) (hf : ∀ a, f (r • a) = r • f a)
+    (M : Matrix m n α) : (r • M).map f = r • M.map f :=
+  ext fun _ _ => hf _
+
+/-- The scalar action via `Mul.toSMul` is transformed by the same map as the elements
+of the matrix, when `f` preserves multiplication. -/
+theorem map_smul' [Mul α] [Mul β] (f : α → β) (r : α) (A : Matrix n n α)
+    (hf : ∀ a₁ a₂, f (a₁ * a₂) = f a₁ * f a₂) : (r • A).map f = f r • A.map f :=
+  ext fun _ _ => hf _ _
+
+/-- The scalar action via `mul.toOppositeSMul` is transformed by the same map as the
+elements of the matrix, when `f` preserves multiplication. -/
+theorem map_op_smul' [Mul α] [Mul β] (f : α → β) (r : α) (A : Matrix n n α)
+    (hf : ∀ a₁ a₂, f (a₁ * a₂) = f a₁ * f a₂) :
+    (MulOpposite.op r • A).map f = MulOpposite.op (f r) • A.map f :=
+  ext fun _ _ => hf _ _
+
+theorem _root_.IsSMulRegular.matrix [SMul R S] {k : R} (hk : IsSMulRegular S k) :
+    IsSMulRegular (Matrix m n S) k :=
+  IsSMulRegular.pi fun _ => IsSMulRegular.pi fun _ => hk
+
+theorem _root_.IsLeftRegular.matrix [Mul α] {k : α} (hk : IsLeftRegular k) :
+    IsSMulRegular (Matrix m n α) k :=
+  hk.isSMulRegular.matrix
+
+instance subsingleton_of_empty_left [IsEmpty m] : Subsingleton (Matrix m n α) :=
+  ⟨fun M N => by
+    ext i
+    exact isEmptyElim i⟩
+
+instance subsingleton_of_empty_right [IsEmpty n] : Subsingleton (Matrix m n α) :=
+  ⟨fun M N => by
+    ext i j
+    exact isEmptyElim j⟩
+
+/-- This is `Matrix.of` bundled as an additive equivalence. -/
+def ofAddEquiv [Add α] : (m → n → α) ≃+ Matrix m n α where
+  __ := of
+  map_add' _ _ := rfl
+
+@[simp] lemma coe_ofAddEquiv [Add α] :
+    ⇑(ofAddEquiv : (m → n → α) ≃+ Matrix m n α) = of := rfl
+@[simp] lemma coe_ofAddEquiv_symm [Add α] :
+    ⇑(ofAddEquiv.symm : Matrix m n α ≃+ (m → n → α)) = of.symm := rfl
+
+end Matrix
+
+open Matrix
+
+namespace Matrix
+
+section Diagonal
```

Comment:
I think I'd argue that `diagonal` and `1` are really multiplicative ideas, and so belong in `Mul.lean` as well.

Reply:
I think `Mul.lean` is getting pretty long as it is, and really what `diagonal`/`1` do is set up the `AddMonoidWithOne` structure (used e.g. in `Data/Matrix/CharP.lean`) rather than the full semiring structure in `Mul.lean`. I'll split `diagonal` off into its own file.